File: 21Everyday/December/yangui/807保持城市天际线/807保持城市天际线.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        lineMax,rowMax = [], []
        for i in range(len(grid)):
            lineMax.append(max(grid[i]))
        for i in range(len(grid)):
            maxItem = 0
            for j in range(len(grid[0])):
                if(grid[j][i] > maxItem):
                    maxItem = grid[j][i]
            rowMax.append(maxItem)
        result = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                result += min(lineMax[i],rowMax[j]) - grid[i][j]
        return result
        
            

File: 21Everyday/December/linhao/807. 保持城市天际线/807. 保持城市天际线.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxIncreaseKeepingSkyline = function (grid) {
  const maxH = [];
  const maxV = [];
  const vgrid = [];
  const cols = grid[0].length;

  for (let n = 0; n < cols; n++) {
    vgrid.push([]);
  }

  for (let i = 0; i < grid.length; i++) {
    const row = grid[i];
    maxH.push(Math.max(...row));

    for (let j = 0; j < row.length; j++) {
      vgrid[j][i] = row[j];
    }
  }

  for (let i = 0; i < vgrid.length; i++) {
    const row = vgrid[i];
    maxV.push(Math.max(...row));
  }

  let ans = 0;
  for (let i = 0; i < grid.length; i++) {
    const row = grid[i];

    for (let j = 0; j < row.length; j++) {
      if (row[j] < maxH[i] && row[j] < maxV[j]) {
        const min = Math.min(maxH[i], maxV[j]);
        ans += min - row[j];
      }
    }
  }

  return ans;
};

const grid = [[3,0,8,4],[2,4,5,7],[9,2,6,3],[0,3,1,0]];

maxIncreaseKeepingSkyline(grid);